### PR TITLE
Ability to quickly discard large compressed events

### PR DESCRIPTION
**Pull request by niemyjski**

### CODE REVIEW

File: src/Exceptionless.Core/Jobs/EventPostsJob.cs [hunk 1]
```diff
@@ -30,47 +30,77 @@ public class EventPostsJob : QueueJobBase<EventPost> {
...
+            _jsonSerializerSettings = jsonSerializerSettings;
 
             AutoComplete = false;
         }
 
         protected override async Task<JobResult> ProcessQueueEntryAsync(QueueEntryContext<EventPost> context) {
             var queueEntry = context.QueueEntry;
-            var eventPostInfo = await _storage.GetEventPostAndSetActiveAsync(queueEntry.Value.FilePath, _logger, context.CancellationToken).AnyContext();
-            if (eventPostInfo == null) {
+            var fileInfo = await _storage.GetFileInfoAsync(queueEntry.Value.FilePath).AnyContext();
```
Comment (niemyjski) [inline src/Exceptionless.Core/Jobs/EventPostsJob.cs]: I need to add a test to Foundatio to see what happens if this path is null or doesn't exist

File: src/Exceptionless.Core/Jobs/EventPostsJob.cs [hunk 2]
```diff
@@ -30,47 +30,77 @@ public class EventPostsJob : QueueJobBase<EventPost> {
...
+            if (fileInfo == null) {
+                await queueEntry.AbandonAsync().AnyContext();
```
Comment (niemyjski) [inline src/Exceptionless.Core/Jobs/EventPostsJob.cs]: If the result is null lets bail quick

File: src/Exceptionless.Core/Jobs/EventPostsJob.cs [hunk 3]
```diff
@@ -30,47 +30,77 @@ public class EventPostsJob : QueueJobBase<EventPost> {
...
+                return JobResult.FailedWithMessage($"Unable to retrieve post data info '{queueEntry.Value.FilePath}'.");
+            }
+
+            if (fileInfo.Size > Settings.Current.MaximumEventPostSize) {
+                await queueEntry.CompleteAsync().AnyContext();
+                return JobResult.FailedWithMessage($"Unable to process post data '{queueEntry.Value.FilePath}' ({fileInfo.Size} bytes): Maximum event post size limit ({Settings.Current.MaximumEventPostSize} bytes) reached.");
```
Comment (niemyjski) [inline src/Exceptionless.Core/Jobs/EventPostsJob.cs]: If it's greater than max size lets not even process it.

File: src/Exceptionless.Core/Jobs/EventPostsJob.cs [hunk 4]
```diff
@@ -30,47 +30,77 @@ public class EventPostsJob : QueueJobBase<EventPost> {
...
+            }
+
+            var ep = await _storage.GetEventPostAndSetActiveAsync(queueEntry.Value.FilePath, _logger, context.CancellationToken).AnyContext();
```
Comment (niemyjski) [inline src/Exceptionless.Core/Jobs/EventPostsJob.cs]: This hits storage again (we added an extra hit just to get metadata above)

File: src/Exceptionless.Core/Jobs/EventPostsJob.cs [hunk 5]
```diff
@@ -30,47 +30,77 @@ public class EventPostsJob : QueueJobBase<EventPost> {
...
+                await CompleteEntryAsync(queueEntry, ep, SystemClock.UtcNow).AnyContext();
                 return JobResult.Success;
             }
 
+            byte[] uncompressedData = ep.Data;
+            if (!String.IsNullOrEmpty(ep.ContentEncoding)) {
+                try {
+                    uncompressedData = await uncompressedData.DecompressAsync(ep.ContentEncoding).AnyContext();
+                } catch (OutOfMemoryException ex) {
+                    await CompleteEntryAsync(queueEntry, ep, SystemClock.UtcNow).AnyContext();
```
Comment (niemyjski) [inline src/Exceptionless.Core/Jobs/EventPostsJob.cs]: Calling `uncompressedMemoryStream.ToArray()` to get the bytes can throw an out of memory exception. IF this is the case bail!

File: src/Exceptionless.Core/Jobs/EventPostsJob.cs [hunk 6]
```diff
@@ -30,47 +30,77 @@ public class EventPostsJob : QueueJobBase<EventPost> {
...
+                    return JobResult.FailedWithMessage($"Unable to decompress post data '{queueEntry.Value.FilePath}' ({ep.Data.Length} bytes compressed): {ex.Message}");
+                }
+
+                long maxUncompressedEventPostSize = Settings.Current.MaximumEventPostSize * 10;
+                if (uncompressedData.Length > maxUncompressedEventPostSize) {
```
Comment (niemyjski) [inline src/Exceptionless.Core/Jobs/EventPostsJob.cs]: I added the check which is 10x the max event post size. We could still get large events through (if it's a single event it could be 10x the limit :\.

File: src/Exceptionless.Core/Jobs/EventPostsJob.cs [hunk 7]
```diff
@@ -138,51 +168,43 @@ public class EventPostsJob : QueueJobBase<EventPost> {
             }
 
             foreach (var requeueEvent in eventsToRetry) {
+                string contentEncoding = null;
+                byte[] data = requeueEvent.GetBytes(_jsonSerializerSettings);
+                if (data.Length > 1000) {
+                    data = await data.CompressAsync().AnyContext();
+                    contentEncoding = "gzip";
+                }
```
Comment (niemyjski) [inline src/Exceptionless.Core/Jobs/EventPostsJob.cs]: @ejsmith I'm wondering if we should even be recompressing it right here.. If it's under our max allowed bytes we probably shouldn't even be dequeueing it.. If we don't compress it here, the check above to get metadata will instantly complete it.

File: src/Exceptionless.Core/Jobs/EventPostsJob.cs [hunk 8]
```diff
@@ -30,47 +30,77 @@ public class EventPostsJob : QueueJobBase<EventPost> {
...
             var queueEntry = context.QueueEntry;
-            var eventPostInfo = await _storage.GetEventPostAndSetActiveAsync(queueEntry.Value.FilePath, _logger, context.CancellationToken).AnyContext();
-            if (eventPostInfo == null) {
+            var fileInfo = await _storage.GetFileInfoAsync(queueEntry.Value.FilePath).AnyContext();
+            if (fileInfo == null) {
+                await queueEntry.AbandonAsync().AnyContext();
+                return JobResult.FailedWithMessage($"Unable to retrieve post data info '{queueEntry.Value.FilePath}'.");
+            }
+
+            if (fileInfo.Size > Settings.Current.MaximumEventPostSize) {
```
Comment (nj) [inline src/Exceptionless.Core/Jobs/EventPostsJob.cs]: Properly a bit redundant, as basically already checked through the `OverageHandler`?

Comment (niemyjski) [inline src/Exceptionless.Core/Jobs/EventPostsJob.cs]: Yes, but it's probably to be safer than sorry, since we are getting massive events in storage. Also overage handler might not work if the content length isn't sent or is malformed??

Comment (nj) [inline src/Exceptionless.Core/Jobs/EventPostsJob.cs]: Yeah, but I guess if ContentLength is not set or even not matching up, there's much worse issues at stake.
